### DISCUSSION TITLE
S2 buffer  Calculation memory is not released

### DISCUSSION THREAD

**Discussion by LemonLiTree** [google/s2geometry#323] [Q&A]
I use S2BufferOperation to calculate a s2point buffer, the code is as follows
```c++
        std::vector<S2Point> points;//here is a S2 point
        MutableS2ShapeIndex shape_index;
        shape_index.Add(std::make_unique<S2PointVectorShape>(points));

        S2BufferOperation::Options options;//the options are set here
        options.set_buffer_radius();
        options.set_circle_segments();
        options.set_end_cap_style();
        options.set_polyline_side();
        auto output = std::make_unique<S2Polygon>();
        S2BufferOperation op(std::make_unique<s2builderutil::S2PolygonLayer>(output.get()), options);

        op.AddShapeIndex(shape_index);

        S2Error error;
        if(!op.Build(&error)){
            //error
            return nullptr;
        }
```
Running this code, I get the correct result.
However, when the amount of data I want to calculate is large, for example, I want to calculate 10,000 S2point buffers, this code will be called 10,000 times in a loop, and the memory usage will be high after running and will not be released.
Does the calculation here do data caching? Is there any way to release the memory in time?

**Comment by LemonLiTree** [comment]
<img width="319" alt="image" src="https://github.com/google/s2geometry/assets/109049295/99166584-968b-422e-ac11-d6c07e40c80a">

**Comment by smcallis** [comment]
Are you re-using the S2BufferOperation object?  Most of the memory used should be in the S2Polygon instance you're passing in, is that not the case?

**Reply by LemonLiTree** [reply]
I put the above code in a function, each time a S2point will be passed in as a parameter to calculate the buffer, but when there are many S2points to be calculated, this function will be called repeatedly

**Reply by LemonLiTree** [reply]
Here should be the problem of my code itself

**Comment by smcallis** [comment]
If it's in a function all those objects should get cleaned up and their memory with them, I'd be surprised if we had a large memory leak in S2BufferOperation.

**Reply by LemonLiTree** [reply]
After my verification, the above code does have the problem of not releasing the memory. At that time, I did not have this problem when debugging on the mac. Only running on linux will have problems, our system uses JEMALLOC for memory management. Can you test it?

**Reply by smcallis** [reply]
Do you know where the memory is being leaked from?

**Reply by LemonLiTree** [reply]
Sorry for the problem with my code, thanks for the reply
